pmpkg-config/pmpkg-config.1
.\" 
.\" pmpkg-config manual page.
.\" Preface and pmpkg-config section (c) 2006 Ivo van Poorten
.\" (C) Red Hat, Inc. based on gnome-config man page (C) Miguel de Icaza (miguel@gnu.org)
.\"
.TH pmpkg-config 1
.SH NAME
pmpkg-config \- Return metainformation about installed libraries
.SH SYNOPSIS
.PP
.B pmpkg-config
[\-\-modversion] [\-\-help] [\-\-print-errors] [\-\-silence-errors] 
[\-\-cflags] [\-\-libs] [\-\-libs-only-L]
[\-\-libs-only-l] [\-\-cflags-only-I]
[\-\-variable=VARIABLENAME]
[\-\-define-variable=VARIABLENAME=VARIABLEVALUE]
[\-\-uninstalled]
[\-\-exists] [\-\-atleast-version=VERSION] [\-\-exact-version=VERSION]
[\-\-max-version=VERSION] [LIBRARIES...]
.SH PREFACE

\fIpmpkg-config\fP stands for Poor Man's pkg-config.
Most of this manual page is a copy of the original \fIpkg-config\fP
manual page.
\fIpmpkg-config\fP is a Bourne Shell script that mimics \fIpkg-config\fP.
Some functionality might be left out though and some incompatibilities
exist.
After reading a certain section below, it is advised to check the
PMPKG-CONFIG NOTES section at the end of this document to see if any of
it applies to the section you just read.
Whether \fIpmpkg-config\fP can be called as \fIpkg-config\fP from the
command line depends on how it was installed by your distribution.

.SH DESCRIPTION

The \fIpkg-config\fP program is used to retrieve information about
installed libraries in the system.  It is typically used to compile
and link against one or more libraries.  Here is a typical usage
scenario in a Makefile:
.PP
.nf
program: program.c
	cc program.c `pkg-config --cflags --libs gnomeui`
.fi
.PP

.PP
\fIpkg-config\fP retrieves information about packages from 
special metadata files. These files are named after the package, 
with the extension \fI.pc\fP. By default, pkg-config looks in 
the directory \fIprefix\fP/lib/pkgconfig for these files; it will also
look in the colon-separated (on Windows, semicolon-separated) 
list of directories specified by the 
PKG_CONFIG_PATH environment variable. 

.PP
The package name specified on the \fIpkg-config\fP command line is
defined to be the name of the metadata file, minus the \fI.pc\fP
extension. If a library can install multiple versions simultaneously,
it must give each version its own name (for example, GTK 1.2 might
have the package name "gtk+" while GTK 2.0 has "gtk+-2.0").

.SH OPTIONS
The following options are supported:
.TP
.I "--modversion"
Requests that the version information of the libraries specified on
the command line be displayed.  If \fIpkg-config\fP can find all the
libraries on the command line, each library's version string is
printed to stdout, one version per line. In this case \fIpkg-config\fP
exits successfully. If one or more libraries is unknown,
\fIpkg-config\fP exits with a nonzero code, and the contents of stdout
are undefined.
.TP
.I "--help"
Displays a help message and terminates.

.TP
.I "--print-errors"
If one or more of the modules on the command line, or their
dependencies, are not found, or if an error occurs in parsing
a \fI.pc\fP file, then this option will cause errors explaining the
problem to be printed. With "predicate" options such as "--exists"
\fIpkg-config\fP runs silently by default, because it's usually used
in scripts that want to control what's output. This option can be used
alone (to just print errors encountered locating modules on the 
command line) or with other options. The PKG_CONFIG_DEBUG_SPEW
environment variable overrides this option.

.TP
.I "--silence-errors"
If one or more of the modules on the command line, or their
dependencies, are not found, or if an error occurs in parsing a
a \fI.pc\fP file, then this option will keep errors explaining the
problem from being printed. With "predicate" options such as
"--exists" \fIpkg-config\fP runs silently by default, because it's
usually used in scripts that want to control what's output. So this
option is only useful with options such as "--cflags" or
"--modversion" that print errors by default. The PKG_CONFIG_DEBUG_SPEW
environment variable overrides this option.

.TP
.I "--errors-to-stdout"
If printing errors, print them to stdout rather than the default stderr

.PP
The following options are used to compile and link programs:
.TP
.I "--cflags"
This prints pre-processor and compile flags required to compile the
packages on the command line, including flags for all their
dependencies. Flags are "compressed" so that each identical flag
appears only once. \fIpkg-config\fP exits with a nonzero code if it
can't find metadata for one or more of the packages on the command
line.
.TP 
.I "--libs"
This option is identical to "--cflags", only it prints the link
flags. As with "--cflags", duplicate flags are merged (maintaining
proper ordering), and flags for dependencies are included in the
output.
.TP
.I "--libs-only-L"
This prints the -L/-R part of "--libs". That is, it defines the 
library search path but doesn't specify which libraries to link with.
.TP
.I "--libs-only-l"
This prints the -l part of "--libs" for the libraries specified on
the command line. Note that the union of "--libs-only-l" and
"--libs-only-L" may be smaller than "--libs", due to flags such as
-rdynamic.

.TP
.I "--variable=VARIABLENAME"
This returns the value of a variable defined in a package's \fI.pc\fP
file. Most packages define the variable "prefix", for example, so you 
can say:
.nf
  $ pkg-config --variable=prefix glib-2.0
  /usr/
.fi
.TP
.I "--define-variable=VARIABLENAME=VARIABLEVALUE"
This sets a global value for a variable, overriding the value in any
\fI.pc\fP files. Most packages define the variable "prefix", for
example, so you can say:
.nf
  $ pkg-config --print-errors --define-variable=prefix=/foo \e
               --variable=prefix glib-2.0
  /foo
.fi

.TP
.I "--uninstalled"
Normally if you request the package "foo" and the package
"foo-uninstalled" exists, \fIpkg-config\fP will prefer the 
"-uninstalled" variant. This allows compilation/linking against
uninstalled packages. If you specify the "--uninstalled" option,
\fIpkg-config\fP will return successfully if any "-uninstalled"
packages are being used, and return failure (false) otherwise.
(The "PKG_CONFIG_DISABLE_UNINSTALLED" environment variable keeps 
\fIpkg-config\fP from implicitly choosing "-uninstalled" packages, so
if that variable is set, they will only have been used if you pass 
a name like "foo-uninstalled" on the command line explicitly.)

.TP
.I "--exists"
.TP
.I "--atleast-version=VERSION"
.TP
.I "--exact-version=VERSION"
.TP
.I "--max-version=VERSION"
These options test whether the package or list of packages on the
command line are known to \fIpkg-config\fP, and optionally 
whether the version number of a package meets certain contraints.
If all packages exist and meet the specified version constraints,
\fIpkg-config\fP exits successfully. Otherwise it exits unsuccessfully.

Rather than using the version-test options, you can simply give a version
constraint after each package name, for example:
.nf
  $ pkg-config --exists 'glib-2.0 >= 1.3.4 libxml = 1.8.3'
.fi
Remember to use \-\-print-errors if you want error messages.

.TP
.I "--msvc-syntax"
This option is available only on Windows. It causes \fIpkg-config\fP
to output -l and -L flags in the form recognized by the Microsoft
Visual C++ command-line compiler, \fIcl\fP. Specifically, instead of
\fI-Lx:/some/path\fP it prints \fI/libpath:x/some/path\fP, and instead
of \fI-lfoo\fP it prints \fIfoo.lib\fP. Note that the --libs output
consists of flags for the linker, and should be placed on the cl
command line after a /link switch. 

.TP
.I "--dont-define-prefix"
This option is available only on Windows. It prevents \fIpkg-config\fP
from automatically trying to override the value of the variable
"prefix" in each .pc file.

.TP
.I "--prefix-variable=PREFIX"
Also this option is available only on Windows. It sets the name of the
variable that \fIpkg-config\fP automatically sets as described above.

.TP
.I "--static"
Output libraries suitable for static linking.  That means including
any private libraries in the output.  This relies on proper tagging in
the .pc files, else a too large number of libraries will ordinarily be
output.

.SH ENVIRONMENT VARIABLES

.TP
.I "PKG_CONFIG_PATH"
A colon-separated (on Windows, semicolon-separated) list of
directories to search for .pc files.  The default directory will
always be searched after searching the path; the default is
\fIlibdir\fP/pkgconfig:\fIdatadir\fP/pkgconfig where \fIlibdir\fP is
the libdir where \fIpkg-config\fP and \fIdatadir\fP is the datadir
where \fIpkg-config\fP was installed.

.TP
.I "PKG_CONFIG_DEBUG_SPEW"
If set, causes \fIpkg-config\fP to print all kinds of
debugging information and report all errors.

.TP
.I "PKG_CONFIG_TOP_BUILD_DIR"
A value to set for the magic variable \fIpc_top_builddir\fP
which may appear in \fI.pc\fP files. If the environment variable is
not set, the default value '$(top_builddir)' will be used. This
variable should refer to the top builddir of the Makefile where the 
compile/link flags reported by \fIpkg-config\fP will be used.
This only matters when compiling/linking against a package that hasn't
yet been installed.

.TP
.I "PKG_CONFIG_DISABLE_UNINSTALLED"
Normally if you request the package "foo" and the package
"foo-uninstalled" exists, \fIpkg-config\fP will prefer the 
"-uninstalled" variant. This allows compilation/linking against
uninstalled packages.  If this environment variable is set, it
disables said behavior.

.TP
.I "PKG_CONFIG_ALLOW_SYSTEM_CFLAGS"
Don't strip -I/usr/include out of cflags.

.TP
.I "PKG_CONFIG_ALLOW_SYSTEM_LIBS"
Don't strip -L/usr/lib out of libs

.TP
.I "PKG_CONFIG_LIBDIR"
Replaces the default \fIpkg-config\fP search directory.

.SH WINDOWS SPECIALITIES
If a .pc file is found in a directory that matches the usual
conventions (i.e., ends with \\lib\\pkgconfig), the prefix for that
package is assumed to be the grandparent of the directory where the
file was found, and the \fIprefix\fP variable is overridden for that
file accordingly.

In addition to the \fIPKG_CONFIG_PATH\fP environment variable, the
Registry keys
\fIHKEY_CURRENT_USER\\Software\\pkgconfig\\PKG_CONFIG_PATH\fP and
\fIHKEY_LOCAL_MACHINE\\Software\\pkgconfig\\PKG_CONFIG_PATH\fP can be
used to specify directories to search for .pc files. Each (string)
value in these keys is treated as a directory where to look for .pc
files.

.SH AUTOCONF MACROS

.TP
.I "PKG_CHECK_MODULES(VARIABLE-PREFIX,MODULES[,ACTION-IF-FOUND,[ACTION-IF-NOT-FOUND]])"

The macro PKG_CHECK_MODULES can be used in \fIconfigure.ac\fP to 
check whether modules exist. A typical usage would be:
.nf
 PKG_CHECK_MODULES([MYSTUFF], [gtk+-2.0 >= 1.3.5 libxml = 1.8.4])
.fi

This would result in MYSTUFF_LIBS and MYSTUFF_CFLAGS substitution
variables, set to the libs and cflags for the given module list. 
If a module is missing or has the wrong version, by default configure
will abort with a message. To replace the default action, 
specify an ACTION-IF-NOT-FOUND. PKG_CHECK_MODULES will not print any
error messages if you specify your own ACTION-IF-NOT-FOUND.
However, it will set the variable MYSTUFF_PKG_ERRORS, which you can 
use to display what went wrong.

.I "PKG_PROG_PKG_CONFIG([MIN-VERSION])"

Defines the PKG_CONFIG variable to the best pkg-config available,
useful if you need pkg-config but don't want to use PKG_CHECK_MODULES.

.SH METADATA FILE SYNTAX
To add a library to the set of packages \fIpkg-config\fP knows about,
simply install a \fI.pc\fP file. You should install this file to 
\fIlibdir\fP/pkgconfig.

.PP
Here is an example file:
.nf
# This is a comment
prefix=/home/hp/unst   # this defines a variable
exec_prefix=${prefix}  # defining another variable in terms of the first
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: GObject                            # human-readable name
Description: Object/type system for GLib # human-readable description
Version: 1.3.1
URL: http://www.gtk.org
Requires: glib-2.0 = 1.3.1
Conflicts: foobar <= 4.5
Libs: -L${libdir} -lgobject-1.3
Libs.private: -lm
Cflags: -I${includedir}/glib-2.0 -I${libdir}/glib/include 
.fi

.PP
You would normally generate the file using configure, of course, so
that the prefix, etc. are set to the proper values.

.PP
Files have two kinds of line: keyword lines start with a keyword plus
a colon, and variable definitions start with an alphanumeric string
plus an equals sign. Keywords are defined in advance and have special
meaning to \fIpkg-config\fP; variables do not, you can have any
variables that you wish (however, users may expect to retrieve the
usual directory name variables).

.PP
Note that variable references are written "${foo}"; you can escape
literal "${" as "$${".

.TP
.I "Name:"
This field should be a human-readable name for the package. Note that
it is not the name passed as an argument to \fIpkg-config\fP.
.TP
.I "Description:"
This should be a brief description of the package
.TP
.I "URL:"
An URL where people can get more information about and download the package
.TP
.I "Version:"
This should be the most-specific-possible package version string.
.TP
.I "Requires:"
This is a comma-separated list of packages that are required by your
package. Flags from dependent packages will be merged in to the flags
reported for your package. Optionally, you can specify the version 
of the required package (using the operators =, <, >, >=, <=);
specifying a version allows \fIpkg-config\fP to perform extra sanity
checks. You may only mention the same package one time on the 
.I "Requires:"
line. If the version of a package is unspecified, any version will
be used with no checking.
.TP
.I "Conflicts:"
This optional line allows \fIpkg-config\fP to perform additional
sanity checks, primarily to detect broken user installations.  The
syntax is the same as
.I "Requires:"
except that
you can list the same package more than once here, for example 
"foobar = 1.2.3, foobar = 1.2.5, foobar >= 1.3", if you have reason to
do so. If a version isn't specified, then your package conflicts with
all versions of the mentioned package. 
If a user tries to use your package and a conflicting package at the
same time, then \fIpkg-config\fP will complain.
.TP
.I "Libs:"
This line should give the link flags specific to your package. 
Don't add any flags for required packages; \fIpkg-config\fP will 
add those automatically.

.TP
.I "Libs.private:"
This line should list any private libraries in use.  Private libraries
are libraries which are not exposed through your library, but are
needed in the case of static linking.

.TP
.I "Cflags:"
This line should list the compile flags specific to your package. 
Don't add any flags for required packages; \fIpkg-config\fP will 
add those automatically.

.SH AUTHOR

\fIpkg-config\fP was written by James Henstridge, rewritten by Martijn
van Beers, and rewritten again by Havoc Pennington. Tim Janik, Owen
Taylor, and Raja Harinath submitted suggestions and some code.
\fIgnome-config\fP was written by Miguel de Icaza, Raja Harinath and
various hackers in the GNOME team.  It was inspired by Owen Taylor's
\fIgtk-config\fP program.

.PP
\fIpmpkg-config\fP was written from scratch by Ivo van Poorten.

.SH BUGS

\fIpkg-config\fP does not handle mixing of parameters with and without
= well.  Stick with one.

.SH PMPKG-CONFIG NOTES

\fIpmpkg-config\fP tries to mimic \fIpkg-config\fP but there are a few
known limitations:

.PP
* It does not work on Windows.
.PP
* It does not allow spaces in variable values when defining variables
on the command line (e.g. \-\-define\-variable=prefix=/usr/My Programs).
.PP
* It does not support uninstalled packages.
.PP
* It does not check the Conflict: fields in .pc files.

.PP
Some of the above might be added in the future.
Or they might not.
Patches and bug-reports can be sent to <mpm-devel@lists.berlios.de>


lazybox/ChangeLog
0.90
- first release


pmpkg-config/pmpkg-config.sh
#! /bin/sh

# Poor Man's Package Config
#
# Copyright (c) 2006 by Ivo van Poorten <ivop@euronet.nl>
# All rights reserved.
#
# License: GNU General Public License version 2.0
# (and NOT at your option any later version)
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# set -v

# do not change the next two lines. a sed script in the Makefile depends on it
MYPREFIX="/usr/local"
VERSION=0.00

test "$#" = "0" && echo "Must specify package names on the command line" &&
    exit 2

case " $@" in
    *\ --help*|*\ -\?)
    cat >&2 << __EOF__
Poor Man's pkg-config $VERSION -- Copyright (c) 2006 by Ivo van Poorten
Usage: $0 [OPTION...]
  --version                               output version of $0
  --modversion                            output version for package
  --atleast-pkgconfig-version=VERSION     require given version of pkg-config
                                          (always exits with 0)
  --libs                                  output all linker flags
  --static                                output linker flags for static
                                          linking
  --short-errors                          print short errors
  --libs-only-l                           output -l flags
  --libs-only-other                       output other libs (e.g. -pthread)
  --libs-only-L                           output -L flags
  --cflags                                output all pre-processor and compiler
                                          flags
  --cflags-only-I                         output -I flags
  --cflags-only-other                     output cflags not covered by the
                                          cflags-only-I option
  --variable=VARIABLENAME                 get the value of a variable
  --define-variable=VARIABLENAME=VALUE    set the value of a variable
  --exists                                return 0 if the module(s) exist
  --atleast-version=VERSION               return 0 if the module is at least
                                          version VERSION
  --exact-version=VERSION                 return 0 if the module is at exactly
                                          version VERSION
  --max-version=VERSION                   return 0 if the module is at most
                                          version VERSION
  --list-all                              list all known packages
  --debug                                 show verbose debug information
  --print-errors                          show verbose information about
                                          missing packages
  --silence-errors                        do not show verbose information
                                          about missing packages
  --errors-to-stdout                      print errors from --print-errors
                                          to stdout and not to stderr
  -?, --help                              show this help message
  --usage                                 display brief usage message
__EOF__
    exit 2 ;;
    *\ --usage*)
    cat >&2 << __EOF__
Usage: $0 [-?] [--version] [--modversion]
        [--atleast-pkgconfig-version=VERSION] [--libs] [--static]
        [--short-errors] [--libs-only-l] [--libs-only-other] [--libs-only-L]
        [--cflags] [--cflags-only-I] [--cflags-only-other]
        [--exists] [--debug] [--help] [--usage] [--print-errors]
        [--silence-errors] [--errors-to-stdout] [--variable=VARIABLENAME]
        [--define-variable=VARIABLENAME=VALUE] [--atleast-version=VERSION]
        [--exact-version=VERSION] [--max-version=VERSION] [--list-all]
__EOF__
    exit 2 ;;
esac

case "$@" in
    *--version*)      echo "$VERSION" ; exit 0 ;;
    *--atleast-pkgconfig-version*)      exit 0 ;;
esac

_plus_private=no
_verbose=normal
_tostdout=no

_listall=no
_exists=no
_cflagsI=no
_cflagso=no
_libsl=no
_libsL=no
_libso=no
_modversion=no
_variable=""
_defenv=""
_defvars=""
_atleastv=""
_exactv=""
_maxv=""

_CFLAGS=""
_LIBSl=""
_LIBSL=""
_MODVERSIONS=""

PKG_CONFIG_LIBDIR=${PKG_CONFIG_LIBDIR:-"$MYPREFIX/lib/pkgconfig"}
PKG_CONFIG_PATH="$PKG_CONFIG_LIBDIR:$PKG_CONFIG_PATH"

# ignore PKG_CONFIG_DEBUG_SPEW as we do not print any debug information
# ignore PKG_CONFIG_TOP_BUILD_DIR as we do not support -uninstalled
# ignore PKG_CONFIG_DISABLE_UNINSTALLED as we do not support -uninstalled

_allow_system_cflags=no
test -n "`set | grep "PKG_CONFIG_ALLOW_SYSTEM_CFLAGS"`" && \
    _allow_system_cflags=yes
_allow_system_libs=no
test -n "`set | grep "PKG_CONFIG_ALLOW_SYSTEM_LIBS"`" && \
    _allow_system_libs=yes

nmod=0

parse_modname() {
    name=`echo $1 | sed 's/,//'`
    eval _mod_$nmod=$name
    if test -n "$2" ; then
        case "$2" in
            \<\=)   c=lteq  ;;
            \<)     c=lt    ;;
            \=)     c=eq    ;;
            \!\=)   c=neq   ;;
            \>)     c=gt    ;;
            \>\=)   c=gteq  ;;
        esac
        eval _mod_${nmod}_c=$c
        eval _mod_${nmod}_v=$3
    else
        eval _mod_${nmod}_c=ok
    fi

    nmod=`expr $nmod + 1`
}

parse_cmd_line() {
    while test "$#" != "0" ; do
        case "$1" in
            --list-all)             _listall=yes ;;
            --exists)               _exists=yes ;;
            --atleast-version=*)    _atleastv=`echo $1 | cut -d '=' -f 2` ;;
            --exact-version=*)      _exactv=`echo $1 | cut -d '=' -f 2` ;;
            --max-version=*)        _maxv=`echo $1 | cut -d '=' -f 2` ;;
            --cflags)               _cflagsI=yes _cflagso=yes ;;
            --cflags-only-I)        _cflagsI=yes _cflagso=no ;;
            --cflags-only-other)    _cflagsI=no  _cflagso=yes ;;
            --libs)                 _libsl=yes _libsL=yes _libso=yes  ;;
            --libs-only-l)          _libsl=yes _libsL=no  _libso=no   ;;
            --libs-only-L)          _libsl=no  _libsL=yes _libso=no  ;;
            --libs-only-other)      _libsl=no  _libsL=no  _libso=yes ;;
            --static)               _plus_private=yes ;;
            --modversion)           _modversion=yes ;;
            --short-errors)         ;;
            --debug)                ;;
            --print-errors)         _verbose=all ;;
            --silence-errors)       _verbose=none ;;
            --errors-to-stdout)     _tostdout=yes ;;
            --variable=*)           _variable=`echo $1 | cut -d '=' -f 2` ;;
            --define-variable=*=*)
                _tmpenv=`echo "$1" | sed -e 's/[^=]*=\(.*\)/\1/'` 
                _defenv="$_defenv $_tmpenv"
                _tmpvar=`echo "$1" | cut -d '=' -f 2`
                _defvars="$_defvars $_tmpvar"
                ;;
            [abcdefghijklmnopqrstuvwxyz0123456789]*)
                case "$2" in
                    \<*|\=*|\>*|\!*)
                        parse_modname $1 $2 $3
                        shift 2
                        ;;
                    *)
                        parse_modname $1
                esac
                ;;
            *)  echo "error parsing command line: $1"; exit 2 ;;
        esac
        shift 1
    done
}

find_file() {
    for i in /usr/lib/pkgconfig /usr/local/lib/pkgconfig `IFS=":"; echo $PKG_CONFIG_PATH` ; do
        test -f $i/$1.pc && echo $i/$1.pc && return
    done
    echo "No package '$1' found" >&2
    echo "notfound"
}

# echo the list. do not echo an entry if it also occurs later in the list
remove_doubles() {
    while test "$#" != "0"; do
        me=$1
        shift 1
        case " $@ " in
            *\ $me\ *)    ;;
            *)          echo "$me" ;;
        esac
    done
}

remove_system() {
    while test "$#" != "0" ; do
        me=$1
        shift 1
        test "$me" = "-I/usr/include" -a "$_allow_system_cflags" != "yes" \
            && continue
        test "$me" = "-L/usr/lib" -a "$_allow_system_libs" != "yes" \
            && continue
        echo $me
    done
}

add_mods() {
    while test "$#" != "0" ; do
        cur=$1
        shift 1
        case "$1" in
            \<*|\=*|\>*|\!*)
                case " $@ " in
                    *\ $cur\ *) ;;
                    *)          parse_modname $cur $1 $2 ;;
                esac
                shift 2
                ;;
            *)
                case " $@ " in
                    *\ $cur\ *) ;;
                    *)          parse_modname $cur ;;
                esac
                ;;
        esac
    done
}

all_requires2() {
    N=0
    while test "$N" != "$nmod" ; do
        mod=`eval echo \\$_mod_$N`
        echo $mod
        file=`find_file $mod`
        if test "$file" = "notfound" ; then echo "notfound"; exit 2; fi
        if test "$_plus_private" = "yes" ; then
            t="`grep '^Requires:' $file | cut -d ':' -f 2` `cat $file | grep '^Requires.private:' | cut -d ':' -f 2`"
#            t=`remove_doubles $t`
        else
            t=`grep '^Requires:' $file | cut -d ':' -f 2`
        fi
        add_mods $t
        N=`expr $N + 1`
    done
}

add_quotes() {
    sed -e 's/\(.*\)=\(.*\)/\1="\2"/'
}

remove_defined() {
    test -z "$_defvar" && cat && exit 0
    for i in $_defvars ; do
        sed -e "/^$i/d"
    done
}

data_from_file() {
    field=$1
    file=$2
    env=`sed -e '/^#.*$/d' -e '/^[ABCDEFGHIJKLMNOPQRSTUVWXYZ][ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_\.]*:/d' $file | remove_defined | add_quotes`
    flags=`cat $file | grep "^$field:" | cut -d ':' -f 2`
    output=`eval "eval $_defenv; $env "; eval echo $flags`
    echo $output
}

var_from_file() {
    var=$1
    file=$2
    env=`sed -e '/^#.*$/d' -e '/^[ABCDEFGHIJKLMNOPQRSTUVWXYZ][ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_\.]*:/d' $file | remove_defined | add_quotes`
    output=`eval "eval $_defenv; $env "; eval echo \$\{$var\}`
    echo $output
}

collect_data() {
    field=$1
    shift 1
    while test "$#" != "0" ; do
        file=`find_file $1`
        data_from_file $field $file
        shift 1
    done
}

extract() {
    which=$1
    shift 1
    while test "$#" != "0" ; do
        case $which in
            L)  case "$1" in
                    \-L*|\-R*)      echo $1 ;;
                esac
                ;;
            l)  case "$1" in
                    \-l*)           echo $1 ;;
                esac
                ;;
            lo) case "$1" in
                    \-l*|\-L*|\-R*)         ;;
                    *)              echo $1 ;;
                esac
                ;;
            D)  case "$1" in
                    \-I*)   ;;
                    *)      echo $1 ;;
                esac
                ;;
            I)  case "$1" in
                    \-I*)   echo $1 ;;
                esac
                ;;
        esac
        shift 1
    done
}

printop() {
    case "$1" in
        lteq)   echo "<="   ;;
        lt)     echo "<"    ;;
        eq)     echo "="    ;;
        neq)    echo "!="   ;;
        gt)     echo ">"    ;;
        gteq)   echo ">="   ;;
    esac
}

# number filename
check_constraint_n() {
    mod=`eval echo \\$_mod_$1`
    c=`eval echo \\$_mod_$1_c`
    v=`eval echo \\$_mod_$1_v`

    if test "$c" = "ok" ; then
        echo ok
        exit 0
    fi

    f=$2

    maj=`echo $v.0.0.0 | cut -d '.' -f 1`
    min=`echo $v.0.0.0 | cut -d '.' -f 2`
    sub=`echo $v.0.0.0 | cut -d '.' -f 3`

    modver=`data_from_file Version $f`
    modmaj=`echo $modver.0.0.0 | cut -d '.' -f 1`
    modmin=`echo $modver.0.0.0 | cut -d '.' -f 2`
    modsub=`echo $modver.0.0.0 | cut -d '.' -f 3`

    newver=`printf '%03d%03d%03d' "$maj" "$min" "$sub"`
    newmodver=`printf '%03d%03d%03d' "$modmaj" "$modmin" "$modsub"`

    errmsg="Requested $mod `printop $c` $v but version of $mod is $modver"
    trap 'echo $errmsg >&2' 0
    case "$c" in
        lteq)   test "$newmodver" -le "$newver" || exit 2 ;;
        lt)     test "$newmodver" -lt "$newver" || exit 2 ;;
        eq)     test "$newmodver" -eq "$newver" || exit 2 ;;
        neq)    test "$newmodver" -ne "$newver" || exit 2 ;;
        gt)     test "$newmodver" -gt "$newver" || exit 2 ;;
        gteq)   test "$newmodver" -ge "$newver" || exit 2 ;;
    esac
    trap '' 0
    echo ok
}

check_all_constraints() {
    n=0
    while test "$n" != "$nmod" ; do
        mod=`eval echo \\$_mod_$n`
        file=`find_file $mod`
        test "$file" != "notfound" || exit 2
        ret=`check_constraint_n $n $file`
        test "$ret" = "ok" || exit 2
        n=`expr $n + 1`
    done
}

parse_cmd_line $@

if test "$_listall" = "yes" ; then
    for i in `IFS=":"; echo $PKG_CONFIG_PATH` ; do
        for j in $i/*.pc ; do
            test -f "$j" || continue
            k=`basename $j .pc`
            m=`grep "^Description:" $j | cut -d ':' -f 2`
            printf '%-19s %s\n' "$k" "$m"
        done
    done
    exit 0
fi

if test "$_tostdout" = "yes" ; then
    exec 2>&1
fi
if test "$_verbose" = "none" ; then
    exec 1>/dev/null 2>&1
fi

if test -n "$_atleastv" -o -n "$_exactv" -o -n "$_maxv" ; then
    _exists=yes
    test -n "$_atleastv" && c=gteq && ver="$_atleastv"
    test -n "$_exactv"   && c=eq   && ver="$_exactv"
    test -n "$_maxv"     && c=lteq && ver="$_maxv"
    n=0
    while test "$n" != "$nmod" ; do
        eval _mod_${n}_c=$c
        eval _mod_${n}_v=$ver
        n=`expr $n + 1`
    done
fi

test "$_cflagsI$_cflagso$_libsl$_libsL$_libso$_modversion" = "nononononono" \
     -a -z "$_variable" && _exists=yes

test "$_cflagsI$_cflagso$_libsl$_libsL$_libso" != "nonononono" && \
    _variable=""

if test "$_exists" = "yes" ; then
    if test "$_verbose" != "all" ; then
        exec 1>/dev/null 2>&1
    fi
    n=0
    while test "$n" != "$nmod" ; do
        mod=`eval echo \\$_mod_$n`
        file=`find_file $mod`
        test "$file" != "notfound" || exit 2
        ret=`check_constraint_n $n $file`
        test "$ret" = "ok" || exit 2
        n=`expr $n + 1`
    done
    exit 0
fi

if test "$_modversion" = "yes" ; then
    n=0
    while test "$n" != "$nmod" ; do
        mod=`eval echo \\$_mod_$n`
        file=`find_file $mod`
        test "$file" != "notfound" || exit 2
        _MODVERSIONS="$_MODVERSIONS `data_from_file Version $file`"
        ret=`check_constraint_n $n $file`
        test "$ret" = "ok" || exit 2
        n=`expr $n + 1`
    done
    for i in $_MODVERSIONS ; do
        echo $i
    done
fi

if test -n "$_variable" ; then
    n=0
    while test "$n" != "$nmod" ; do
        mod=`eval echo \\$_mod_$n`
        file=`find_file $mod`
        test "$file" != "notfound" || exit 2
        _VARIABLES="$_VARIABLES `var_from_file $_variable $file`"
        ret=`check_constraint_n $n $file`
        test "$ret" = "ok" || exit 2
        n=`expr $n + 1`
    done
    echo $_VARIABLES
    exit 0
fi

check_all_constraints || exit 2
requires="`all_requires2`"
case "$requires" in
    *notfound*)     exit 2 ;;
esac
check_all_constraints || exit 2
requires=`remove_doubles $requires`
if test "$_cflagsI" = "yes" -o "$_cflagso" = "yes"; then
    _CFLAGS=`collect_data Cflags $requires`
    _CFLAGS=`remove_doubles $_CFLAGS`
    _CFLAGS=`remove_system $_CFLAGS`
    test "$_cflagsI" = "yes" && _CFLAGSI=`extract I $_CFLAGS`
    test "$_cflagso" = "yes" && _CFLAGSo=`extract D $_CFLAGS`
fi
if test "$_libsl" = "yes" -o "$_libsL" = "yes" -o "$_libso" = "yes" ; then
    _LIBS=`collect_data Libs $requires`
    if test "$_plus_private" = "yes"; then
        _LIBSP=`collect_data Libs.private $requires`
    else
        _LIBSP=""
    fi
    _LIBS="$_LIBS $_LIBSP"
    _LIBS=`remove_doubles $_LIBS`
    _LIBS=`remove_system $_LIBS`
    test "$_libsL" = "yes" && _LIBSL=`extract L $_LIBS`
    test "$_libsl" = "yes" && _LIBSl=`extract l $_LIBS`
    test "$_libso" = "yes" && _LIBSo=`extract lo $_LIBS`
fi

PRINT=no
LINE=""
test "$_cflagsI" = "yes" && LINE="$LINE $_CFLAGSI" && PRINT=yes
test "$_cflagso" = "yes" && LINE="$LINE $_CFLAGSo" && PRINT=yes
test "$_libsL" = "yes" && LINE="$LINE $_LIBSL" && PRINT=yes
test "$_libsl" = "yes" && LINE="$LINE $_LIBSl" && PRINT=yes
test "$_libso" = "yes" && LINE="$LINE $_LIBSo" && PRINT=yes
test "$PRINT" = "yes" && echo $LINE

exit 0
